bundles/org.openhab.binding.viessmann/src/main/java/org/openhab/binding/viessmann/internal/handler/ViessmannBridgeHandler.java

```diff
@@ -0,0 +1,572 @@
+        this.linkRegistry = linkRegistry;
+    }
+
+    @Override
+    public void setInstallationGatewayId(@Nullable String newInstallation, @Nullable String newGateway) {
+        newInstallationId = newInstallation;
+        newGatewaySerial = newGateway;
+    }
+
+    /**
+     * get the devices list (needed for discovery)
+     *
+     * @return a list of the all devices
+     */
+    public List<String> getDevicesList() {
+        // return a copy of the list, so we don't run into concurrency problems
+        return new ArrayList<>(devicesList);
+    }
+
+    public List<Thing> getChildren() {
+        return getThing().getThings().stream().filter(Thing::isEnabled).collect(Collectors.toList());
+    }
+
+    private void setConfigInstallationGatewayId() {
+        Configuration conf = editConfiguration();
+        conf.put("installationId", newInstallationId);
+        conf.put("gatewaySerial", newGatewaySerial);
+        updateConfiguration(conf);
+    }
+
+    private boolean errorChannelsLinked() {
+        return getThing().getChannels().stream()
+                .anyMatch(c -> isLinked(c.getUID()) && ERROR_CHANNELS.contains(c.getUID().getId()));
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (channelUID.getId().equals(CHANNEL_RUN_QUERY_ONCE) && OnOffType.ON.equals(command)) {
+            logger.debug("Received command: CHANNEL_RUN_QUERY_ONCE");
+            pollingFeatures();
+            updateState(CHANNEL_RUN_QUERY_ONCE, OnOffType.OFF);
+        }
+        if (channelUID.getId().equals(CHANNEL_RUN_ERROR_QUERY_ONCE) && OnOffType.ON.equals(command)) {
+            logger.debug("Received command: CHANNEL_RUN_ERROR_QUERY_ONCE");
+            getDeviceError();
+            updateState(CHANNEL_RUN_ERROR_QUERY_ONCE, OnOffType.OFF);
+        }
+    }
+
+    @Override
+    public void dispose() {
+        stopViessmannBridgePolling();
+        stopViessmannErrorsPolling();
+        stopViessmannBridgeLimitReset();
+    }
+
+    @Override
+    public Collection<Class<? extends ThingHandlerService>> getServices() {
+        return Set.of(ViessmannDiscoveryService.class);
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initialize Viessmann Accountservice");
+
+        BridgeConfiguration config = getConfigAs(BridgeConfiguration.class);
+        this.config = config;
+        String storedApiCalls = this.stateStorage.get(STORED_API_CALLS);
+        if (storedApiCalls != null) {
+            apiCalls = Integer.parseInt(storedApiCalls);
+        } else {
+            apiCalls = 0;
+        }
+        newInstallationId = "";
+        newGatewaySerial = "";
+
+        api = new ViessmannApi(this.config.apiKey, httpClient, this.config.user, this.config.password,
+                this.config.installationId, this.config.gatewaySerial, callbackUrl);
+        api.createOAuthClientService(this);
+
+        if (api.doAuthorize()) {
```

Comment:
these are blocking calls and need to be async. Similar to the accounthandler's init